### PR TITLE
Add SwitchProducer mechanism to allow runtime decision which algorithm implementation to run

### PR DESCRIPTION
**Pull request by makortel**
This PR provides an implementation for a `SwitchProducer` to allow a runtime decision on which EDProducer from a given set to run for a given module label. The main use case in mind are heterogeneous/offloaded/accelerated algorithms, where we would run the accelerated EDProducer if the accelerator device is available in the system, and run the "legacy" CPU EDProducer otherwise.

For a given accelerator type, the generic `SwitchProducer` is supposed to be inherited along
```python
def _switch_foo():
    return (isFooDeviceEnabled(), 2)

class SwitchProducerFoo(cms.SwitchProducer):
    def __init__(self, **kargs):
        super(SwitchProducerFoo,self).__init__(
            dict(cpu = SwitchProducer.getCpu(),
                 foo = _switch_foo),
            **kargs
        )
```
Here the deriving class provides a dictionary of possible cases along with functions that return a `(bool, int)` tuple. The `bool` tells whether that device type is enabled on the machine, and the `int` gives a priority of that device wrt. other devices. `SwitchProducer.getCpu()` returns a function for the CPU (always enabled, priority `1`). Note that in order to be able to pickle the configuration, these functions have to be at (python) module level (i.e. e.g. lambdas or static methods are not sufficient).

The decision logic chooses the case that has the largest priority among those that are enabled. The decision is made at the point where the python configuration is transformed for C++, i.e. in case of production at the worker node.

The switch decision does not affect the configuration hash, meaning that the framework will consider a run/lumi/event produced with `isFooDeviceEnabled() == False` the same as with `True`.

The concrete `SwitchProducerFoo` would then be used along (e.g. in a `cfi` file)
```python
bar = SwitchProducerFoo(
    cpu = cms.EDProducer("Bar", ...),
    foo = cms.EDProducer("BarOnFooDevice", ...)
)
```
In this example, if `isFooDeviceEnabled()` returns
* `True` => `BarOnFooDevice` EDProducer is run
* `False` => `Bar` EDProducer is run

The switched EDProducers are required to declare exactly the same products with the `produces()` call.

When a client asks for a products if `bar`, it will get pointed to the exactly the same `edm::Wrapper` that contains the product by the chosen EDProducer. The additional "layer' is, however, tracked in provenance, such that the parent of a product `bar` is `bar@cpu` or `bar@foo`, depending which one of the EDProducers was run.

The python interface is likely to evolve with more experience with it, with accelerated EDProducers, and with devices beyond CUDA. Note that this PR adds only the generic infrastructure, a CUDA-specific implementation will follow in a separate PR (or e.g. as part of the effort of #25353).

Tested in 10_4_0_pre3, no changes expected.

### CODE REVIEW

**Comment by fwyzard** [comment]
> The switched EDProducers are required to declare exactly the same products with the produces() call.

Doesn't this go in the opposite direction of what we discussed earlier, with the possibility of having different "chain" of modules used for e.g. the CPU vs GPU reconstruction ?

Or do you expect the switch to be used only for the last module in the chain, and relay on the dependencies to trigger different chains ?

But, wouldn't that lead to different provenances ?

**Comment by makortel** [comment]
@cmsbuild, please test

File: DataFormats/Common/src/ProductData.cc [hunk 1]
```diff
@@ -38,4 +38,8 @@ namespace edm {
   void ProductData::unsafe_setWrapper(std::unique_ptr<WrapperBase> iValue) const {
     wrapper_ = std::move(iValue);
   }
+
+  void ProductData::unsafe_setWrapper(std::shared_ptr<WrapperBase const> iValue) const {
+    wrapper_ = std::const_pointer_cast<WrapperBase>(std::move(iValue));
```
Comment (Dr15Jones) [inline DataFormats/Common/src/ProductData.cc]: Is it possible to just make `wrapper_` a `std::shared_ptr<const WrapperBase>` ?

File: FWCore/Framework/interface/ProductRegistryHelper.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ namespace edm {
       TypeID typeID_;
       std::string productInstanceName_;
       std::string branchAlias_;
+      std::string switchAlias_;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/ProductRegistryHelper.h]: Do we really need two strings? Can they both ever be set? Maybe just an enum to say what type of alias? Could hid access to it behind a member function.

File: FWCore/Framework/src/Principal.cc [hunk 3]
```diff
@@ -171,9 +171,20 @@ namespace edm {
     if(hasAliases) {
       for(auto const& prod : prodsList) {
         BranchDescription const& bd = prod.second;
-        if(bd.isAlias() && bd.branchType() == branchType_) {
+        if(bd.isAnyAlias() && bd.branchType() == branchType_) {
           auto cbd = std::make_shared<BranchDescription const>(bd);
-          addAliasedProduct(cbd);
+          if(bd.isSwitchAlias()) {
+            if(bd.onDemand()) {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/Principal.cc]: Maybe some documentation here explaining the reason?

File: FWCore/Framework/test/stubs/ToyIntProducers.cc [hunk 4]
```diff
@@ -411,6 +412,39 @@ namespace edmtest {
...
+
+  class ManyIntProducer : public edm::global::EDProducer<> {
+  public:
+    explicit ManyIntProducer(edm::ParameterSet const& p):
+      tokenValues_{vector_transform(p.getParameter<std::vector<edm::ParameterSet>>("values"), [this](edm::ParameterSet const& pset) {
+          return TokenValue{produces<IntProduct>(pset.getParameter<std::string>("instance")),
+                            pset.getParameter<int>("value")};
+        })}
+    {
+      tokenValues_.push_back(TokenValue{produces<IntProduct>(), p.getParameter<int>("ivalue")});
```
Comment (Dr15Jones) [inline FWCore/Framework/test/stubs/ToyIntProducers.cc]:
```suggestion
      tokenValues_.emplace_back(produces<IntProduct>(), p.getParameter<int>("ivalue"));
```

File: FWCore/Integration/test/SwitchProducerProvenanceAnalyzer.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    edm::Handle<IntProduct> h;
+    iEvent.getByToken(inputToken1_, h);
+    testProduct(h, iEvent.id().luminosityBlock(), iEvent);
+    
+    iEvent.getByToken(inputToken2_, h);
+    testProduct(h, iEvent.id().luminosityBlock(), iEvent);
+  }
+
+  void SwitchProducerProvenanceAnalyzer::testProduct(edm::Handle<IntProduct> const& prod, int mode, edm::Event const& iEvent) const {
+    assert(prod->value == mode);
```
Comment (Dr15Jones) [inline FWCore/Integration/test/SwitchProducerProvenanceAnalyzer.cc]: Do other test modules use assert or do they use exceptions?

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

**Comment by Dr15Jones** [comment]
I know we have a lot of them, but I'm not a big fan of tests which are checked for success based on diffs with a log file. I much prefer them to fail outright with checks being done by other modules in the job.

File: FWCore/Framework/test/stubs/ToyIntProducers.cc (hunk 4 above)
Comment (makortel) [reply FWCore/Framework/test/stubs/ToyIntProducers.cc]: That would require the `TokenValue` struct to have a constructor. I can of course do that if you prefer, I went this way to save some repeated typing.

Comment (makortel) [reply FWCore/Framework/test/stubs/ToyIntProducers.cc]: Doesn't work unless I add a

File: DataFormats/Common/src/ProductData.cc (hunk 1 above)
Comment (makortel) [reply DataFormats/Common/src/ProductData.cc]: In principle yes, I suppose. But I believe that would mean adding `const_cast` to
https://github.com/cms-sw/cmssw/blob/a480cc2041dd20d6ebd2ec67eb050b055f90f728/DataFormats/Common/interface/ProductData.h#L36
, or deal with "modifying a `const` object" in
https://github.com/cms-sw/cmssw/blob/a480cc2041dd20d6ebd2ec67eb050b055f90f728/FWCore/Framework/src/ProductResolvers.cc#L93-L110
or
https://github.com/cms-sw/cmssw/blob/a480cc2041dd20d6ebd2ec67eb050b055f90f728/DataFormats/Common/interface/WrapperDetail.h#L98-L102

**Comment by makortel** [comment]
> I know we have a lot of them, but I'm not a big fan of tests which are checked for success based on diffs with a log file. I much prefer them to fail outright with checks being done by other modules in the job.

I don't have strong feelings for the diffs, so I could add an analyzer to the test configurations that would throw with wrong input.

File: FWCore/Framework/src/Principal.cc (hunk 3 above)
Comment (makortel) [reply FWCore/Framework/src/Principal.cc]: For the differentiation of on-demand and non-on-demand `SwitchProducer`? Will add.

**Comment by makortel** [comment]
Comments addressed

**Comment by fabiocos** [comment]
@makortel in my opinion squashing during the review might help in preserving the subsequent comment history (of course the previous one is lost), it would make sense to do it after the reviewer has acknowledged the updates, @Dr15Jones do you agree?

**Comment by Dr15Jones** [comment]
lets go ahead and squash now.

**Comment by makortel** [comment]
Tidied up the history. Apparently nowadays GitHub provides a nice link to show the diff of a force-push. That provides a way to see a diff of an update even if some earlier commits got updated (as long as the base commit stays the same).

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
+1

**Pull request merged**